In [2]:
#Importing libraries
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn import metrics as mt
import statsmodels.formula.api as smf
import statsmodels.api as sm


In [3]:
#Importing database
df = pd.read_csv("database/train.csv")
df.head()

id_cliente  idade  saldo_atual  divida_atual  renda_anual  \
0        1767     21   278.172008       2577.05  24196.89636   
1       11920     40   268.874152       2465.39  19227.37796   
2        8910     36   446.643127       1055.29  42822.28223   
3        4964     58   321.141267        703.05  51786.82600   
4       10100     35   428.716114        891.29  44626.85346   

   valor_em_investimentos  taxa_utilizacao_credito  num_emprestimos  \
0              104.306544                31.038763                6   
1               69.858778                36.917093                5   
2              134.201478                34.561714                0   
3              297.350067                31.493561                0   
4              134.201478                28.028887                2   

   num_contas_bancarias  num_cartoes_credito  dias_atraso_dt_venc  \
0                     5                    7                   21   
1                     8                    5                   40   
2                     3                    6                   26   
3                     3                    7                   12   
4                     8                    7                   24   

   num_pgtos_atrasados  num_consultas_credito  taxa_juros investe_exterior  \
0                   14                      9          15              Não   
1                   23                     10          18              Não   
2                   13                      3          15              Sim   
3                    7                      2           1              Sim   
4                   10                      8          20              Sim   

  pessoa_polit_exp limite_adicional  
0              Não            Negar  
1              Não            Negar  
2              Não            Negar  
3              Não            Negar  
4              Não            Negar

# Data Preparation

In [4]:
features = ['idade', 'divida_atual', 'renda_anual',
       'valor_em_investimentos', 'taxa_utilizacao_credito', 'num_emprestimos',
       'num_contas_bancarias', 'num_cartoes_credito', 'dias_atraso_dt_venc',
       'num_pgtos_atrasados', 'num_consultas_credito', 'taxa_juros']
label = 'saldo_atual'

In [5]:
x_train = df.loc[:, features]
y_train = df.loc[:, label]

# Model Training

In [6]:
lr_model = LinearRegression()
lr_model.fit(x_train, y_train)

LinearRegression()

In [7]:
y_pred = lr_model.predict(x_train)
y_pred

array([346.66954862, 367.84027655, 431.46897895, ..., 412.25174802,
       400.68529863, 455.26482252])

In [8]:
df1 = df.copy()
df1 = df1.loc[:, ['id_cliente', 'saldo_atual']]
df1['pred'] = y_pred
df1.head()

id_cliente  saldo_atual        pred
0        1767   278.172008  346.669549
1       11920   268.874152  367.840277
2        8910   446.643127  431.468979
3        4964   321.141267  445.506463
4       10100   428.716114  378.271169

# Performance

In [9]:
mt.r2_score(y_train, y_pred)

0.16917364489050013

# Model training statsmodels

In [10]:
lr_model = reg2 = smf.ols(formula='saldo_atual ~ idade + divida_atual + num_emprestimos', data = df)
lr_model = lr_model.fit()
anova_results2 = sm.stats.anova_lm(lr_model, typ=1)
anova_results2

df        sum_sq       mean_sq            F  \
idade               1.0  4.085925e+02  4.085925e+02     0.009575   
divida_atual        1.0  4.494005e+07  4.494005e+07  1053.104993   
num_emprestimos     1.0  3.976109e+03  3.976109e+03     0.093174   
Residual         9496.0  4.052309e+08  4.267385e+04          NaN   

                        PR(>F)  
idade             9.220529e-01  
divida_atual     3.537135e-219  
num_emprestimos   7.601864e-01  
Residual                   NaN

In [11]:
#mean square error - mede o erro quadrado, principal problema é que da alta importancia aos outliers
mse = np.round(mt.mean_squared_error(y_train, y_pred), 2)
print(mse)

39370.27


In [12]:
#root mean square error - a raiz quadrada do erro quadratico, volta para o valor original (a mesma escala da variavel alvo), para meio de comparações

rmse = np.round(mt.root_mean_squared_error(y_train, y_pred), 2)
print(rmse)

198.42


In [13]:
df.describe()

id_cliente        idade  saldo_atual  divida_atual   renda_anual  \
count   9500.000000  9500.000000  9500.000000   9500.000000  9.500000e+03   
mean    6210.532947   126.398421   404.532032   1437.086420  1.950373e+05   
std     3603.243914   685.036005   217.696725   1156.263217  1.454274e+06   
min        1.000000    18.000000     0.088628      0.230000  1.047387e+04   
25%     3091.750000    28.000000   272.363060    573.620000  2.620790e+04   
50%     6172.500000    38.000000   337.169588   1175.730000  4.599111e+04   
75%     9320.250000    47.000000   471.264657   1963.332500  8.044983e+04   
max    12500.000000  8598.000000  1602.040519   4998.070000  2.327966e+07   

       valor_em_investimentos  taxa_utilizacao_credito  num_emprestimos  \
count             9500.000000              9500.000000      9500.000000   
mean               611.999323                32.270823        10.863684   
std               2002.443880                 5.106053        62.555417   
min                  0.000000                20.992914         0.000000   
25%                 76.083981                28.078872         2.000000   
50%                134.201478                32.262649         3.000000   
75%                249.182915                36.464575         6.000000   
max              10000.000000                49.564519      1496.000000   

       num_contas_bancarias  num_cartoes_credito  dias_atraso_dt_venc  \
count           9500.000000          9500.000000          9500.000000   
mean              17.552316            21.433158            20.990316   
std              120.980186           122.677390            14.800612   
min                0.000000             1.000000            -5.000000   
25%                3.000000             4.000000            10.000000   
50%                6.000000             6.000000            18.000000   
75%                7.000000             7.000000            28.000000   
max             1779.000000          1479.000000            67.000000   

       num_pgtos_atrasados  num_consultas_credito   taxa_juros  
count          9500.000000            9500.000000  9500.000000  
mean             29.832421              24.937158    66.618632  
std             222.637763             183.733346   440.118524  
min               0.000000               0.000000     1.000000  
25%               9.000000               2.000000     8.000000  
50%              14.000000               4.000000    14.000000  
75%              18.000000               8.000000    20.000000  
max            4388.000000            2589.000000  5788.000000

In [25]:
#Criando um outlier em apenas 1 resultado na coluna saldo atual e refazendo o rmse

indice_aleatorio = df['saldo_atual'].sample(475).index[0]

# Multiplicar o valor correspondente por 10 e atualizar no DataFrame
df.at[indice_aleatorio, 'saldo_atual'] *= 100

In [26]:
lr_model2 = LinearRegression()
lr_model2.fit(x_train, y_train)
y_pred = lr_model.predict(x_train)
y_pred

0       336.783121
1       343.434261
2       427.354210
3       448.303860
4       437.087727
           ...    
9495    479.688981
9496    442.191232
9497    356.230181
9498    460.030506
9499    439.491648
Length: 9500, dtype: float64

In [27]:
rmse = np.round(mt.root_mean_squared_error(y_train, y_pred), 2)
print(rmse)

739.85


Com 5% de valores outlier o rmse de 198.42 foi para 739.85, em um df com varios outliers teremos um rmse com um valor discrepante